File: LeetCode/1510. Stone Game IV.java
/***  KNOWLEDGE IS POWER  ***/

class Solution {
    public boolean winnerSquareGame(int n) {
        boolean dp[] = new boolean[n + 1];
        for(int i = 1; i <= n; i++)
        {
            for(int j = 1; j * j <= i; j++)
            {
                if(!dp[i - j * j])
                {
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[n];
    }
}

File: LeetCode/1510. Stone Game IV.py
###   KNOWLEDGE IS POWER  ###

class Solution:
    @cache
    def winnerSquareGame(self, n: int) -> bool:
        if not n: 
            return False
        for i in reversed(range(1, floor(sqrt(n)) + 1)):
            if not self.winnerSquareGame(n - i**2): 
                return True
        return False
